### PR TITLE
hubble: Add node name filter

### PR DESCRIPTION
**Pull request by twpayne**
This is a WIP PR working towards. https://github.com/cilium/hubble/issues/266, and is at a suitable stage for initial review input.

It currently:
* Implements mini glob-like pattern-matching language for node names (so you can write patterns like `test-cluster/k8s*`).
* Adds node name patterns to `flowpb.FlowFilter`.
* Supports white- and blacklisting of node name patterns.
* Filters the list of peers in Hubble Relay to only retrieve flows from matching nodes.

To do:
* [x] Implement server-side filtering on the nodes (i.e. if node name does not match the filter, then never return flows).
* [x] Document node name pattern syntax.
* [ ] Add `--node-name` option to Hubble CLI (https://github.com/cilium/hubble/pull/412)
* [ ] Probably other stuff I missed.

### CODE REVIEW

File: pkg/hubble/filters/node.go
Comment (glibsm) [inline pkg/hubble/filters/node.go:15]: some of the files are missing headers

File: pkg/hubble/observer/local_observer.go
Comment (gandro) [inline pkg/hubble/observer/local_observer.go]: Please use `nodeTypes.GetName()` here. It is the value that we assign to the `node_name` field in the protobuf responses.

https://github.com/cilium/cilium/blob/3533fd35792566f06f52bf9d11e7fb9150233d7d/pkg/node/types/nodename.go#L38-L41

(unrelated to your PR, but something I noticed during the review: We should include the cluster name in the nodename. I have opened #13872 for that. But again, unrelated to your PR, using `nodeTypes.GetName()` is enough to be consistent with the content of the `NodeName` field for now)

**Review by gandro — COMMENTED** [review]
Very solid and clean implementation, awesome work! I found one minor issue with the node name used, see inline comment.

File: pkg/hubble/filters/node.go
Comment (twpayne) [reply pkg/hubble/filters/node.go:15]: Fixed, thanks.

File: pkg/hubble/observer/local_observer.go
Comment (twpayne) [reply pkg/hubble/observer/local_observer.go]: Done, thanks.

**Comment by twpayne** [comment]
test-me-please

**Review by michi-covalent — COMMENTED** [review]
thanks @twpayne the pr looks good to me overall 🚀 2 things i want to get feedback from the community:

- it might be a bit cleaner to have a separate filter for cluster names. we did something similar for `namespace/podname` filter, and it got pretty confusing.
- personally i'm fine with either regex or glob. if we are going with glob, we could probably use something like https://golang.org/pkg/path/filepath/#Match

**Comment by twpayne** [comment]
> * personally i'm fine with either regex or glob. if we are going with glob, we could probably use something like https://golang.org/pkg/path/filepath/#Match

Initially I wanted to use exactly this function, but there are several problems:
* Importantly, with current versions of Go,`path/filepath.Match` evaluates the pattern lazily. That means that there is no way to reliably detect an invalid pattern. This was a problem for many years and has only been fixed less than two weeks ago (https://github.com/golang/go/issues/28614) and will only be available with Go 1.16.
* `path/filepath.Match` behaves differently on Windows systems. Of course we have no plans to run on Windows.
* `path/filepath.Match` is designed to match on filepaths. Kubernetes node names are not filepaths, and `filepath.Match`'s patterns are not a great fit for them.

Regular expressions are quite a heavy solution where it is likely that only straightforward matching will be required. For example, with a regular expression you have to escape every `.` in the domain name to avoid matching any character.

**Comment by gandro** [comment]
> it might be a bit cleaner to have a separate filter for cluster names. we did something similar for `namespace/podname` filter, and it got pretty confusing.

This is actually a very good point @michi-covalent! But I think this implementation here does not suffer from the same shortcomings as the pod filter currently: It allows for `clustername/*` and  `*/nodename` and patterns separately (I think, correct  me if I'm wrong @twpayne). Especially `*/podname` is not supported  by the pod filter currently, which I think is one of the main reasons it's currently hard to use.

**Review by gandro — APPROVED** [review]

**Comment by twpayne** [comment]
> It allows for `clustername/*` and `*/nodename` and patterns separately (I think, correct me if I'm wrong @twpayne)

This is correct. Note that a single `*` currently matches a single domain name component, so if your node names are `cluster-name/foo.bar.baz` you'll have to write `cluster-name/*.*.*` to match them.

I was thinking of adding a double star pattern `**` which matches any number of domain name components so you could write `cluster-name/**` instead. What do you think?

**Comment by gandro** [comment]
> This is correct. Note that a single `*` currently matches a single domain name component, so if your node names are `cluster-name/foo.bar.baz` you'll have to write `cluster-name/*.*.*` to match them.
> 
> I was thinking of adding a double star pattern `**` which matches any number of domain name components so you could write `cluster-name/**` instead. What do you think?

Ah right. So yes, I think we either want `**` or as @michi-covalent suggested, have two separate filters.

I'm impartial to either solution, but it seems to me that being able to include/exclude nodes from a specific cluster regardless of the number of dots in the node name should be a supported case.

**Comment by twpayne** [comment]
OK, PR updated:
* Add `**` pattern to match one or more domain name component so you can write `test-cluster/**` to match all nodes in `test-cluster`.
* Use non-capturing groups in the internal regexps. Capturing groups aren't used anyway and this makes it a little more explicit.

File: pkg/hubble/filters/node.go
Comment (sayboras) [inline pkg/hubble/filters/node.go]: componets -> components

Comment (ti-mo) [inline pkg/hubble/filters/node.go]:
```suggestion
// or more domain name components. All other characters are invalid.
```

File: pkg/hubble/relay/observer/server.go
Comment (sayboras) [inline pkg/hubble/relay/observer/server.go:85]: nit: as peers can be empty here, we can return fast here if you want

File: pkg/hubble/filters/node.go
Comment (ti-mo) [inline pkg/hubble/filters/node.go:86]: Would it be preferable to use the standard `^` and `$` to denote SOL and EOL respectively? I'm not very familiar with the Go regexp implementation and had to look up why `\A` and `\z` are used here, so TIL. Are they functionally different from each other? As far as I can see, `\A`/`\z` don't support line-based matching and only operate on the full string.

File: pkg/hubble/observer/local_observer.go
Comment (sayboras) [inline pkg/hubble/observer/local_observer.go:260]: whitelist and blacklist might be useful info for debugging.

**Review by sayboras — APPROVED** [review]

File: pkg/hubble/relay/observer/server.go
Comment (ti-mo) [inline pkg/hubble/relay/observer/server.go]: My initial read of this confused me, as it sounds a bit contradictory. The final sentence should mention that the backing array is re-used on top of a new slice being returned.

File: api/v1/flow/flow.proto
Comment (ti-mo) [inline api/v1/flow/flow.proto]: There's an incoming change claiming the same proto field here: https://github.com/cilium/cilium/pull/13826/files#diff-74cced194f034ecadb39f852432c1822b6ef022065df0cb0d2ff6c718b339d45R380.

One of these PRs will require a rebase. cc @nyrahul

**Review by ti-mo — CHANGES_REQUESTED** [review]
:100: very easy to read. :+1: Appreciate the fact that all internal methods are documented!

**Review by michi-covalent — APPROVED** [review]
lgtm 🚀 should be good to merge once remaining comments get addressed.

personally i feel it's slightly simpler to have a separate cluster_name filter but `cluster-name/**` works too 👍

File: pkg/hubble/filters/node.go
Comment (twpayne) [reply pkg/hubble/filters/node.go]: Thanks, fixed.

File: pkg/hubble/relay/observer/server.go
Comment (twpayne) [reply pkg/hubble/relay/observer/server.go:85]: I want to leave this as-is for the moment, as the list of peers will change dynamically in the future (#12833).

File: pkg/hubble/observer/local_observer.go
Comment (twpayne) [reply pkg/hubble/observer/local_observer.go:260]: Good idea. However, the blacklist and whitelist will already be printed in the log message printed by the `defer` statement in the previous block of code.

File: pkg/hubble/filters/node.go
Comment (twpayne) [reply pkg/hubble/filters/node.go:86]: Your understanding is correct. I'm using `\A` and `\z` here as they are stricter than `^` and `$`. Node names should never have newlines in them, but we still want to make sure that we're matching on whole strings, not lines within a string.

File: pkg/hubble/relay/observer/server.go
Comment (twpayne) [reply pkg/hubble/relay/observer/server.go]: Thanks, I've updated the function's comment to hopefully be clearer.

File: api/v1/flow/flow.proto
Comment (twpayne) [reply api/v1/flow/flow.proto]: Thanks for the heads-up. I'm happy to rebase this one if #13826 lands first. I'm pretty sure the ProtoBuf compiler will print an error if we accidentally use the same field number twice.

File: pkg/hubble/filters/node.go
Comment (ti-mo) [reply pkg/hubble/filters/node.go:86]: Alright! Doesn't look like we need to worry about that unless multiline support is enabled explicitly, though. (like `"(?m)^.*$"`) IMO it would be clearer with the standard syntax, but it's :fine: having it like this. :+1:

**Comment by twpayne** [comment]
@ti-mo Do you want any more changes? If not, could you approve this PR or at least remove the "requested changes" review?

**Review by ti-mo — APPROVED** [review]

File: pkg/hubble/relay/observer/server.go
Comment (rolinh) [inline pkg/hubble/relay/observer/server.go:85]: I am a little confused about the reason to filter the list of peers to which the request gets forwarded to based on the node name filters in Hubble Relay. Could you elaborate on this please?

File: api/v1/flow/flow.proto
Comment (rolinh) [inline api/v1/flow/flow.proto]: Every other field name in `FlowFilter` is not pluralized although they're all repeated fields. Please, change to `node_name` for consistency.

**Review by rolinh — CHANGES_REQUESTED** [review]
Regarding `clustername/nodename`, I think [this issue](https://github.com/cilium/hubble/issues/359) which summarizes the current problems with the `namespace/podname` filter is worth a read (especially [this comment](https://github.com/cilium/hubble/issues/359#issuecomment-702202613)).

What I am concerned about is that whatever we end up doing, we're consistent across the board. For instance, in the comment linked above one of the proposal is to that `/podname` translates to "all pods named `podname` in every namespaces" and `nsname/` means "all pods in namespace `nsname`" (similarly, a single `/` means "all pods in all namespaces"). This makes a ton of sense to me and is more clear than if it had to be specified as `**/podname` etc.

In light of this, I would suggest to use a similar pattern instead of the double star `**`.  This means that to filter for every node in a cluster named `foo`, the filter would be `foo/`. Similarly, to get all nodes named `bar` in every cluster, one would use `/bar`. And all nodes in all clusters would be `/` although this is probably not a very useful filter.

I think that adding a new `cluster_name` filter as suggested by @michi-covalent would not allow certain types of filters that using patterns in `node_name` allows.

TLDR: before merging this, I want to make sure we think carefully about the implications of adding new pattern types for filtering.

**Comment by michi-covalent** [comment]
@Rolinh hmm yeah that's a good point, i agree in general it's a good idea to keep things consistent.

talking about consistency, there is `dns_query` filter that does something similar as well https://github.com/cilium/cilium/blob/76996fb03e63d8b201dc72374bb05103fb199d43/pkg/hubble/filters/fqdn.go#L40 🤔 

just to satisfy my intellectual curiosity, what kind of filters cannot be supported by having separate filters for cluster_name and node_name?

File: api/v1/flow/flow.proto
Comment (twpayne) [reply api/v1/flow/flow.proto]: Done.

**Comment by twpayne** [comment]
> just to satisfy my intellectual curiosity, what kind of filters cannot be supported by having separate filters for cluster_name and node_name?

@michi-covalent Admittedly this is a rather contrived example, but there are filters that cannot be supported by having separate filters. Consider the following set of clusters and nodes:

```
cluster1/node1
cluster1/node2
cluster2/node1
cluster2/node2
```

The unrepresentable filter is "only traffic from `cluster1/node1` to `cluster2/node2`".

With separate filters, you can use `node_name = ["cluster1/node1", "cluster2/node2"]`. With a single filter you need `cluster_name = ["cluster1", "cluster2"]` and `node_name = ["node1", "node2"]` which matches all traffic between all nodes and clusters, not just `cluster1/node1` to `cluster2/node2`.

File: pkg/hubble/relay/observer/server.go
Comment (twpayne) [reply pkg/hubble/relay/observer/server.go:85]: Ah, this looks like a misunderstanding on my part, thanks for catching this.

The goal was to avoid opening connections to peers that will never generate traffic that matches the filter. I was making the assumption that peers and nodes are the same thing, but of course that is not the case: peers can forward flows from other peers and other nodes.

Refactor coming up...

**Comment by twpayne** [comment]
Thanks for all the reviews, there are two key problems with this PR in its current state I that I need to address:
* Flow filtering should be done on the `node_name` field in the flow, not on the node name of the node running the code.
* The pattern syntax needs to be updated for consistency with other parts of the code.

I'll close this PR for now and re-open it once these issues are addressed.

**Comment by sayboras** [comment]
You can also convert it to draft PR, so that notification will not be sent (I believe :pray: )

**Comment by michi-covalent** [comment]
@twpayne yeah this turned out to be a bit more involved than i initially thought, thanks for this PR though 👍 . i think we have a much better understanding about how we want to approach it. let's follow up either on slack or sig-hubble meeting.

as for your example, you can provide multiple filters and they get ORed, so i believe:
```
{"node_name": ["cluster1/node1", "cluster2/node2"]}
```
is equivalent to:
```
{ "cluster_name": "cluster1", "node_name": "node1"}
{ "cluster_name": "cluster2", "node_name": "node2"}
```

**Pull request closed without merging**
